File: faketokens.py
#!/usr/bin/env python3
"""
Test Token Generator

Generates fake tokens for testing and development purposes.
These tokens are NOT real and cannot be used for actual authentication.

Supported token formats:
- GitHub Classic: ghp_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX (40 chars)
- GitHub Fine-grained: github_pat_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX (varies)
- GitLab: glpat-XXXXXXXXXXXXXXXXXXXX (26 chars total)
- NPM: npm_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX (40 chars)
- AWS Access Key: AKIA + 16 chars (20 chars total)
- AWS Secret Key: 40 character base64-like string
"""

import argparse
import random
import string
import secrets
import sys
from typing import List


class TestTokenGenerator:
    """Generate fake tokens for testing purposes."""
    
    # Token prefixes
    GITHUB_CLASSIC_PREFIX = "ghp_"
    GITHUB_FINE_GRAINED_PREFIX = "github_pat_"
    GITLAB_PREFIX = "glpat-"
    NPM_PREFIX = "npm_"
    AWS_ACCESS_KEY_PREFIX = "AKIA"
    
    # Characters used in tokens
    TOKEN_CHARS = string.ascii_letters + string.digits
    NPM_TOKEN_CHARS = string.ascii_letters + string.digits + "_-"  # NPM allows underscores and hyphens
    AWS_SECRET_CHARS = string.ascii_letters + string.digits + "+/"  # Base64-like characters
    
    def __init__(self):
        """Initialize the token generator."""
        self.generated_tokens = set()
    
    def generate_github_classic_token(self, ensure_unique: bool = True) -> str:
        """
        Generate a fake classic GitHub personal access token.
        
        Args:
            ensure_unique: If True, ensures the token hasn't been generated before
            
        Returns:
            A fake token string starting with 'ghp_'
        """
        while True:
            # Classic tokens have 36 characters after the prefix
            token_body = ''.join(secrets.choice(self.TOKEN_CHARS) for _ in range(36))
            token = f"{self.GITHUB_CLASSIC_PREFIX}{token_body}"
            
            if not ensure_unique or token not in self.generated_tokens:
                if ensure_unique:
                    self.generated_tokens.add(token)
                return token
    
    def generate_github_fine_grained_token(self, length: int = 93, ensure_unique: bool = True) -> str:
        """
        Generate a fake fine-grained GitHub personal access token.
        
        Args:
            length: Total length of the token (default 93 is common)
            ensure_unique: If True, ensures the token hasn't been generated before
            
        Returns:
            A fake token starting with 'github_pat_'
        """
        if length < len(self.GITHUB_FINE_GRAINED_PREFIX) + 10:
            raise ValueError(f"Token length must be at least {len(self.GITHUB_FINE_GRAINED_PREFIX) + 10}")
        
        while True:
            # Calculate body length after removing prefix
            body_length = length - len(self.GITHUB_FINE_GRAINED_PREFIX)
            token_body = ''.join(secrets.choice(self.TOKEN_CHARS) for _ in range(body_length))
            token = f"{self.GITHUB_FINE_GRAINED_PREFIX}{token_body}"
            
            if not ensure_unique or token not in self.generated_tokens:
                if ensure_unique:
                    self.generated_tokens.add(token)
                return token
    
    def generate_gitlab_token(self, ensure_unique: bool = True) -> str:
        """
        Generate a fake GitLab personal access token.
        
        Args:
            ensure_unique: If True, ensures the token hasn't been generated before
            
        Returns:
            A fake token string starting with 'glpat-'
        """
        while True:
            # GitLab tokens have 20 characters after the prefix (26 total)
            token_body = ''.join(secrets.choice(self.TOKEN_CHARS) for _ in range(20))
            token = f"{self.GITLAB_PREFIX}{token_body}"
            
            if not ensure_unique or token not in self.generated_tokens:
                if ensure_unique:
                    self.generated_tokens.add(token)
                return token
    
    def generate_aws_access_key(self, ensure_unique: bool = True) -> str:
        """
        Generate a fake AWS access key ID.
        
        Args:
            ensure_unique: If True, ensures the token hasn't been generated before
            
        Returns:
            A fake AWS access key starting with 'AKIA'
        """
        while True:
            # AWS access keys have 16 characters after AKIA prefix (20 total)
            token_body = ''.join(secrets.choice(string.ascii_uppercase + string.digits) for _ in range(16))
            token = f"{self.AWS_ACCESS_KEY_PREFIX}{token_body}"
            
            if not ensure_unique or token not in self.generated_tokens:
                if ensure_unique:
                    self.generated_tokens.add(token)
                return token
    
    def generate_aws_secret_key(self, ensure_unique: bool = True) -> str:
        """
        Generate a fake AWS secret access key.
        
        Args:
            ensure_unique: If True, ensures the token hasn't been generated before
            
        Returns:
            A fake AWS secret key (40 characters, base64-like)
        """
        while True:
            # AWS secret keys are 40 characters, base64-like encoding
            token = ''.join(secrets.choice(self.AWS_SECRET_CHARS) for _ in range(40))
            
            if not ensure_unique or token not in self.generated_tokens:
                if ensure_unique:
                    self.generated_tokens.add(token)
                return token
    
    def generate_npm_token(self, ensure_unique: bool = True) -> str:
        """
        Generate a fake NPM access token.
        
        Args:
            ensure_unique: If True, ensures the token hasn't been generated before
            
        Returns:
            A fake token string starting with 'npm_'
        """
        while True:
            # NPM tokens have 36 characters after the prefix
            token_body = ''.join(secrets.choice(self.NPM_TOKEN_CHARS) for _ in range(36))
            token = f"{self.NPM_PREFIX}{token_body}"
            
            if not ensure_unique or token not in self.generated_tokens:
                if ensure_unique:
                    self.generated_tokens.add(token)
                return token
    
    def generate_batch(self, count: int, token_type: str = "github_classic") -> List[str]:
        """
        Generate multiple test tokens at once.
        
        Args:
            count: Number of tokens to generate
            token_type: Type of token ("github_classic", "github_fine_grained", "gitlab", "npm", "aws_access_key", or "aws_secret_key")
            
        Returns:
            List of generated tokens
        """
        tokens = []
        for _ in range(count):
            if token_type.lower() == "github_classic":
                tokens.append(self.generate_github_classic_token())
            elif token_type.lower() == "github_fine_grained":
                tokens.append(self.generate_github_fine_grained_token())
            elif token_type.lower() == "gitlab":
                tokens.append(self.generate_gitlab_token())
            elif token_type.lower() == "npm":
                tokens.append(self.generate_npm_token())
            elif token_type.lower() == "aws_access_key":
                tokens.append(self.generate_aws_access_key())
            elif token_type.lower() == "aws_secret_key":
                tokens.append(self.generate_aws_secret_key())
            else:
                raise ValueError("token_type must be 'github_classic', 'github_fine_grained', 'gitlab', 'npm', 'aws_access_key', or 'aws_secret_key'")
        
        return tokens
    
    def clear_history(self):
        """Clear the history of generated tokens."""
        self.generated_tokens.clear()
    
    def get_generated_count(self) -> int:
        """Get the number of unique tokens generated so far."""
        return len(self.generated_tokens)


def main():
    """Main function with command line argument parsing."""
    parser = argparse.ArgumentParser(
        description="Generate fake tokens for testing (GitHub, GitLab, NPM, and AWS)",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  %(prog)s                          # Generate 1 GitHub classic token
  %(prog)s -c 5                     # Generate 5 GitHub classic tokens
  %(prog)s -t github_fine_grained   # Generate GitHub fine-grained token
  %(prog)s -t gitlab                # Generate GitLab token
  %(prog)s -t npm                   # Generate NPM token
  %(prog)s -t aws_access_key        # Generate AWS access key
  %(prog)s -t aws_secret_key        # Generate AWS secret key
  %(prog)s -c 3 -t aws_access_key   # Generate 3 AWS access keys
        """
    )
    
    parser.add_argument(
        "-c", "--count",
        type=int,
        default=1,
        help="Number of tokens to generate (default: 1)"
    )
    
    parser.add_argument(
        "-t", "--type",
        choices=["github_classic", "github_fine_grained", "gitlab", "npm", "aws_access_key", "aws_secret_key"],
        default="github_classic",
        help="Token type to generate (default: github_classic)"
    )
    
    args = parser.parse_args()
    
    generator = TestTokenGenerator()
    
    # Generate and output tokens
    tokens = generator.generate_batch(args.count, args.type)
    for token in tokens:
        print(token)


if __name__ == "__main__":
    main()

File: faketokens.js
#!/usr/bin/env node

/**
 * Test Token Generator
 * 
 * Generates fake tokens for testing and development purposes.
 * These tokens are NOT real and cannot be used for actual authentication.
 * 
 * Supported token formats:
 * - GitHub Classic: ghp_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX (40 chars)
 * - GitHub Fine-grained: github_pat_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX (varies)
 * - GitLab: glpat-XXXXXXXXXXXXXXXXXXXX (26 chars total)
 * - NPM: npm_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX (40 chars)
 * - AWS Access Key: AKIA + 16 chars (20 chars total)
 * - AWS Secret Key: 40 character base64-like string
 */

const crypto = require('crypto');

class TestTokenGenerator {
    constructor() {
        this.generatedTokens = new Set();
        
        // Token prefixes
        this.GITHUB_CLASSIC_PREFIX = "ghp_";
        this.GITHUB_FINE_GRAINED_PREFIX = "github_pat_";
        this.GITLAB_PREFIX = "glpat-";
        this.NPM_PREFIX = "npm_";
        this.AWS_ACCESS_KEY_PREFIX = "AKIA";
        
        // Character sets
        this.TOKEN_CHARS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
        this.NPM_TOKEN_CHARS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_-";
        this.AWS_SECRET_CHARS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+/";
        this.AWS_ACCESS_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    }

    /**
     * Generate random string from character set
     * @param {string} chars - Character set to use
     * @param {number} length - Length of string to generate
     * @returns {string} Random string
     */
    generateRandomString(chars, length) {
        let result = '';
        for (let i = 0; i < length; i++) {
            const randomIndex = crypto.randomInt(0, chars.length);
            result += chars[randomIndex];
        }
        return result;
    }

    /**
     * Generate a fake classic GitHub personal access token
     * @param {boolean} ensureUnique - If true, ensures the token hasn't been generated before
     * @returns {string} A fake token string starting with 'ghp_'
     */
    generateGitHubClassicToken(ensureUnique = true) {
        while (true) {
            const tokenBody = this.generateRandomString(this.TOKEN_CHARS, 36);
            const token = this.GITHUB_CLASSIC_PREFIX + tokenBody;
            
            if (!ensureUnique || !this.generatedTokens.has(token)) {
                if (ensureUnique) {
                    this.generatedTokens.add(token);
                }
                return token;
            }
        }
    }

    /**
     * Generate a fake fine-grained GitHub personal access token
     * @param {number} length - Total length of the token (default 93)
     * @param {boolean} ensureUnique - If true, ensures the token hasn't been generated before
     * @returns {string} A fake token starting with 'github_pat_'
     */
    generateGitHubFineGrainedToken(length = 93, ensureUnique = true) {
        if (length < this.GITHUB_FINE_GRAINED_PREFIX.length + 10) {
            throw new Error(`Token length must be at least ${this.GITHUB_FINE_GRAINED_PREFIX.length + 10}`);
        }
        
        while (true) {
            const bodyLength = length - this.GITHUB_FINE_GRAINED_PREFIX.length;
            const tokenBody = this.generateRandomString(this.TOKEN_CHARS, bodyLength);
            const token = this.GITHUB_FINE_GRAINED_PREFIX + tokenBody;
            
            if (!ensureUnique || !this.generatedTokens.has(token)) {
                if (ensureUnique) {
                    this.generatedTokens.add(token);
                }
                return token;
            }
        }
    }

    /**
     * Generate a fake GitLab personal access token
     * @param {boolean} ensureUnique - If true, ensures the token hasn't been generated before
     * @returns {string} A fake token string starting with 'glpat-'
     */
    generateGitLabToken(ensureUnique = true) {
        while (true) {
            const tokenBody = this.generateRandomString(this.TOKEN_CHARS, 20);
            const token = this.GITLAB_PREFIX + tokenBody;
            
            if (!ensureUnique || !this.generatedTokens.has(token)) {
                if (ensureUnique) {
                    this.generatedTokens.add(token);
                }
                return token;
            }
        }
    }

    /**
     * Generate a fake NPM access token
     * @param {boolean} ensureUnique - If true, ensures the token hasn't been generated before
     * @returns {string} A fake token string starting with 'npm_'
     */
    generateNpmToken(ensureUnique = true) {
        while (true) {
            const tokenBody = this.generateRandomString(this.NPM_TOKEN_CHARS, 36);
            const token = this.NPM_PREFIX + tokenBody;
            
            if (!ensureUnique || !this.generatedTokens.has(token)) {
                if (ensureUnique) {
                    this.generatedTokens.add(token);
                }
                return token;
            }
        }
    }

    /**
     * Generate a fake AWS access key ID
     * @param {boolean} ensureUnique - If true, ensures the token hasn't been generated before
     * @returns {string} A fake AWS access key starting with 'AKIA'
     */
    generateAwsAccessKey(ensureUnique = true) {
        while (true) {
            const tokenBody = this.generateRandomString(this.AWS_ACCESS_CHARS, 16);
            const token = this.AWS_ACCESS_KEY_PREFIX + tokenBody;
            
            if (!ensureUnique || !this.generatedTokens.has(token)) {
                if (ensureUnique) {
                    this.generatedTokens.add(token);
                }
                return token;
            }
        }
    }

    /**
     * Generate a fake AWS secret access key
     * @param {boolean} ensureUnique - If true, ensures the token hasn't been generated before
     * @returns {string} A fake AWS secret key (40 characters, base64-like)
     */
    generateAwsSecretKey(ensureUnique = true) {
        while (true) {
            const token = this.generateRandomString(this.AWS_SECRET_CHARS, 40);
            
            if (!ensureUnique || !this.generatedTokens.has(token)) {
                if (ensureUnique) {
                    this.generatedTokens.add(token);
                }
                return token;
            }
        }
    }

    /**
     * Generate multiple test tokens at once
     * @param {number} count - Number of tokens to generate
     * @param {string} tokenType - Type of token to generate
     * @returns {string[]} Array of generated tokens
     */
    generateBatch(count, tokenType = "github_classic") {
        const tokens = [];
        
        for (let i = 0; i < count; i++) {
            switch (tokenType.toLowerCase()) {
                case "github_classic":
                    tokens.push(this.generateGitHubClassicToken());
                    break;
                case "github_fine_grained":
                    tokens.push(this.generateGitHubFineGrainedToken());
                    break;
                case "gitlab":
                    tokens.push(this.generateGitLabToken());
                    break;
                case "npm":
                    tokens.push(this.generateNpmToken());
                    break;
                case "aws_access_key":
                    tokens.push(this.generateAwsAccessKey());
                    break;
                case "aws_secret_key":
                    tokens.push(this.generateAwsSecretKey());
                    break;
                default:
                    throw new Error("token_type must be 'github_classic', 'github_fine_grained', 'gitlab', 'npm', 'aws_access_key', or 'aws_secret_key'");
            }
        }
        
        return tokens;
    }

    /**
     * Clear the history of generated tokens
     */
    clearHistory() {
        this.generatedTokens.clear();
    }

    /**
     * Get the number of unique tokens generated so far
     * @returns {number} Number of unique tokens generated
     */
    getGeneratedCount() {
        return this.generatedTokens.size;
    }
}

/**
 * Parse command line arguments
 * @returns {object} Parsed arguments
 */
function parseArgs() {
    const args = process.argv.slice(2);
    const result = {
        count: 1,
        type: "github_classic",
        help: false
    };

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        
        if (arg === '-h' || arg === '--help') {
            result.help = true;
        } else if (arg === '-c' || arg === '--count') {
            if (i + 1 >= args.length) {
                throw new Error(`${arg} requires a value`);
            }
            result.count = parseInt(args[++i]);
            if (isNaN(result.count) || result.count < 1) {
                throw new Error('Count must be a positive integer');
            }
        } else if (arg === '-t' || arg === '--type') {
            if (i + 1 >= args.length) {
                throw new Error(`${arg} requires a value`);
            }
            const validTypes = ["github_classic", "github_fine_grained", "gitlab", "npm", "aws_access_key", "aws_secret_key"];
            result.type = args[++i];
            if (!validTypes.includes(result.type)) {
                throw new Error(`Invalid token type. Choose from: ${validTypes.join(', ')}`);
            }
        } else {
            throw new Error(`Unknown argument: ${arg}`);
        }
    }

    return result;
}

/**
 * Display help message
 */
function showHelp() {
    console.log(`Test Token Generator

Generate fake tokens for testing (GitHub, GitLab, NPM, and AWS)

Usage: node faketokens.js [options]

Options:
  -c, --count <number>    Number of tokens to generate (default: 1)
  -t, --type <type>       Token type to generate (default: github_classic)
  -h, --help             Show this help message

Token types:
  github_classic         GitHub classic tokens (ghp_...)
  github_fine_grained    GitHub fine-grained tokens (github_pat_...)
  gitlab                 GitLab tokens (glpat-...)
  npm                    NPM tokens (npm_...)
  aws_access_key         AWS access keys (AKIA...)
  aws_secret_key         AWS secret keys

Examples:
  node faketokens.js                          # Generate 1 GitHub classic token
  node faketokens.js -c 5                     # Generate 5 GitHub classic tokens
  node faketokens.js -t github_fine_grained   # Generate GitHub fine-grained token
  node faketokens.js -t gitlab                # Generate GitLab token
  node faketokens.js -t npm                   # Generate NPM token
  node faketokens.js -t aws_access_key        # Generate AWS access key
  node faketokens.js -t aws_secret_key        # Generate AWS secret key
  node faketokens.js -c 3 -t aws_access_key   # Generate 3 AWS access keys`);
}

/**
 * Main function
 */
function main() {
    try {
        const args = parseArgs();
        
        if (args.help) {
            showHelp();
            return;
        }

        const generator = new TestTokenGenerator();
        const tokens = generator.generateBatch(args.count, args.type);
        
        tokens.forEach(token => console.log(token));
        
    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

// Run if this file is executed directly
if (require.main === module) {
    main();
}

module.exports = TestTokenGenerator;
